readme.txt
How to compile
--------------

Congratulations, you checked out the simutrans source. To compile it,
you have two options, either using Microsoft Visual C++ Express (which
is free in Version 7.0 or up) or GCC.

To compile you will need the following libraries:
libz (http://www.zlib.net/)
libpng (http://www.libpng.org/pub/png/) for makeobj
libbz2.lib (compile from source from http://www.bzip.org/downloads.html)

For the recommended SDL-support you need
libSDL (http://www.libsdl.org/)
libSDL_mixer (link from the same page)

The link for allegro lib is:
http://www.talula.demon.co.uk/allegro/
or
http://alleg.sourceforge.net/index.de.html

To make life easier, you can follow the instructions to compile OpenTTD:
http://wiki.openttd.org/Category:Compiling_OpenTTD
A system set up for OpenTTD will also compile simutrans (except for
bzlib2, see below sections).

If you are on a MS Windows machine, download either MS VC Express or
MingW. The latter is easier to use as part of the DEV-C++ IDE, which
makes the installation of additional libraries like libz and libsdl and so
on very easy. However, to compile the command line is easier.

For all other systems, it is recommended you get latest GCC 3.46 or higher
and matching zlib, libbzip2, and libpng and a game library. For linux
systems you may have to use tools like apt-get or yast2.

To built on Haiku you must use GCC4 (type "setarch x86" in the current
nightlies). To incorporate bz2lib, download make bz2lib and add them
manually (via FLAGS = -I/dwonloadeddir -L/downloadeddir).

A subversion will be also a good idea. You can find some of them on:
http://subversion.tigris.org/
or you some other client.

Check out the latest source from the SVN or check out a certain revision.
I recommend always to use the latest source, since it does not make any
sense to work with buggy code.

The address is:
svn://tron.yamagi.org/simutrans
username is "anon"!

A commandline would look like this:
svn checkout svn://tron.yamagi.org/simutrans --username=anon


IMPORTANT:
----------

If you want to contribute, read the coding guidelines in
sim/documentation/coding_styles.txt


The following instructions are for GCC systems:
-----------------------------------------------

Go to Simutrans/sim.

Then copy the file sim/config.template to sim/config.default and edit the
file. You need to specify:
- frontend (gdi, allegro, sdl)
- color depth (usually 16)
- system (you should know it)

I recommend to uncomment #DEBUG=1 and #OPTIMISE = 1 (i.e. removing the #).

For allegro or libsdl you may need to define the path of the config file
(or at least on win98 and empty path).

Finally type make. If you want a smaller program and do not care about error
messages, you can comment out #DEBUG=1 and run strip sim resp. strip sim.exe
after compile and linking.

For users on window systems:
To debug, I recommend to run drmingw -i once in a shell. You will get a
caller history in case of an error. gdb does not really work well and is a
pain to use with the text interface.


The following instructions are for MS Visual C Express:
-------------------------------------------------------

Download Visual Express C++ (tested for 2008)
http://www.microsoft.com/express/Downloads/

In the OpenTTD wiki you will also find some useful instruction on
setting up MSVC:
http://wiki.openttd.org/Microsoft_Visual_C%2B%2B_2008_Express_Editions

You will need some libraries like zlib etc. Any version which is
compatible with VC will do. Take a look at the instructions for
compiling OpenTTD, which contains a libpng and a zlib suited for
simutrans too:
http://binaries.openttd.org/extra/openttd-useful/

The bzip2 source tarball comes with an archive where you can easily built
your own libbz2.lib file. Or use the one posted in the forum:
http://forum.simutrans.com/index.php?topic=652.msg37080#msg37080

For debugging, you have to set the correct working directory, i.e. the
directory where the pak/ folders are located and use the -use_workdir
command line option.

Berlin, Mai 2011


simutrans/history.txt
    ADD: Map number is shown in the enlarge map window.
    FIX: Text area GUI components no longer suffer from line wrap around when displaying an extreme number of lines.
    FIX: Stop information window waiting list in "via (amount)" order will no longer count incorrectly when more than 8,388,607 units are headed to a single destination.
    FIX: SDL2 build running on Windows 10 in fullscreen mode now minimizes and maximizes responsively.
	CHG: gui overhaul: automatic placement and size calculation of elements
	ADD: use animated water graphics on shores
	FIX: Do not call DefWindowsProc for IME to mitigate crash on Windows1809
	ADD: always use temporary save name *.sv_ during saving before deleting main file
	CHG: Make double pak warning translatable
	FIX: (ACarlotti) rotation of traffic light phases and overflow was broken for years
	FIX: fields now check for allowed climate, not same height
	ADD: (HyperSim inspired) new tool for merging distance stations (and lot of code cleanup of old merging tool)
	ADD: constrait[prev][0]=any require a vehile in front (and same for [next] to require a trailing vehicle)
	CHG: The acceleration tool now also alters the acceleration in fast forward mode
	ADD: Up to 3x3 city buildings, all with up to 8 rotations


Release of 120.4.1 (r8600 on 23-Sep-2018):
	FIX: wrong graphics with greedy block signals
	FIX: world minimum height could not be set
	FIX: location forest could not work with factories smaller than 3x3, since not enough inner tiles were iterated 


Release of 120.4 (r8588 on 17-Sep-2018):
	FIX: properly handle dualstack servers by using both IPv4 and IP6 addresses (if no symbolic name is available)
	ADD: (HyperSim) sorting vehicles in depot
	ADD: system clipboard support now also with SDL2
	FIX: underground slice view now correctly displays front tile walls when a tile infront is flat and 2 elevations below the underground slice
	OPT: significantly improved the performance of changing underground mode/slice when playing on very large maps
	ADD: More factory locations (shore, river, forest) to the existing city, water, land
	ADD: (THLeaderH) option to hide the revenue income messaging at stops
	ADD: (THLeaderH) configurable height setting for extremely heigh maps (use at own risk)
	ADD: (gauthier) "priority" signals (greedy multi block signals)
	FIX: different obj (ways, bridges, tunnels, ... ) with same name failed for checksum comparision


Release of 120.3 (r8503 on 15-June-2018):
	ADD: new blue-white theme
	ADD: new general tool 40 which rotates buildings
	ADD: scenario_info now dialog tool nummer 31
	ADD: Support Truetype fonts on windows up to 19 pts. For MAC and Linux support there is a way needed to find the font path.
	ADD: Server can be now started at the UI when loading a game.
	ADD: switch -easyserver on commandline to host a game with dynamic IP behind a router
	FIX: crash when resizing isometric minimap
	FIX: problems with scrolled lists
	ADD: (Ichou) force replacement of wayobj by holding ctrl key while building
	ADD: (THLeader) long convoi patch
	FIX: removed line management window memory leaks
	ADD: animated pedestrians, parameters:
		steps_per_frame (internal steps per frame, straight movement across tile is 256 steps)
		image[dir][a] (dir - direction, a - animation counter, a=0,1,..)
	CHG: pedestrian movement, new dat parameter:
		offset (of pedestrian image relative to right tile boundary, 0..255, default 20, 0 = on the right edge of tile, 128 = on center line of tile
	FIX: corrected container UI component focus logic (fixes http://forum.simutrans.com/index.php?topic=17277 )
	FIX: added redraw logic for field removal (fixes http://forum.simutrans.com/index.php?topic=17308 )
	FIX: added Unicode file path support for Windows builds
	FIX: removed Windows builds file search resource leak
	FIX: improved Windows build clipboard Unicode parsing
	ADD: sorting in depot (HyperSim)
	FIX: factories which gain or lose fields should no longer have a chance to break with JIT2 (fixes https://forum.simutrans.com/index.php?topic=17394 )
	ADD: sync step barrier for multiplayer clients to prevent them running ahead of the server in the case of an unreliable connection
	CHG: various multiplayer related settings have been given more reliable default values
	FIX: fullscreen Windows GDI no longer alters display area when alt+tab cycled
	ADD: JIT2 demand based production rate scaling
	CHG: changed factory information view to be easier to understand, especially when operating with JIT2
	CHG: speed can be directly selected in goods dialogue
	FIX: consistent and wider climate borders, also other waterlevels than -2 when loading a relief map


Release of 120.2.2: (r8163 on 31-3-2017):
	ADD: Name filter in depot window
	FIX: Combo boxes now work much more reliable
	ADD: Shift+waybuilder on elevated ways selects the tile under the way (for upgrades)
	FIX: Crash when extending convoy length
	ADD: All configurable colors now rgb internally

Release of 120.2.1: (r8121 on 26-Feb-2017):
	FIX: SDL2 with non-ASCII input
	FIX: crashes with SDL2 when resizing
	FIX: allow docks only next to water
	CHG: use almost no assembler any more, the compiler can do better than us

Release of 120.2: (r8077 on 12-Feb-2016):
	FIX: sorting according to displayed name in house/cur/industry building lists
	ADD: scriptable AI players (folder AI which contains sqai now with distribution)
	ADD: (jk271) switch to load heightmaps with more height levels
	FIX: reload themes after restart
	FIX: demo.sve were never loaded, since the auto-load assumed existence even if there was none
	ADD/FIX: slopes for groundobj now working again. If there are groundobjs with climate water, then these will be used occasionally at the coastline
	ADD: ground name=shore with Image[slope][0,1=snow, if there] for display at shoreline
	FIX: smoke tile when rotated factory views
	ADD: length of max tiles checker after choose sign/signal configurable (max_choose_route_steps)
	CHG: update to squirrel 3.1

simobj.h
/*
 * Copyright (c) 1997 - 2001 Hj. Malthaner
 *
 * This file is part of the Simutrans project under the artistic license.
 * (see license.txt)
 */

#ifndef simobjs_h
#define simobjs_h

#include "simtypes.h"
#include "display/clip_num.h"
#include "display/simimg.h"
#include "simcolor.h"
#include "dataobj/koord3d.h"


class cbuffer_t;
class karte_ptr_t;
class player_t;

/**
 * Base class of all objects on the map, obj == thing
 * Since everything is a 'obj' on the map, we need to make this as compact and fast as possible.
 *
 * @author Hj. Malthaner
 */
class obj_t
{
public:
	// flags
	enum flag_values {
		no_flags=0,  /// no special properties
		dirty=1,        /// mark image dirty when drawing
		not_on_map=2,   /// this object is not placed on any tile (e.g. vehicles in a depot)
		is_vehicle=4,   /// this object is a vehicle obviously
		highlight=8      /// for drawing some highlighted outline
	};

	// display only outline with player color on owner stuff
	static bool show_owner;

private:
	obj_t(obj_t const&);
	obj_t& operator=(obj_t const&);

	/**
	 * Coordinate of position
	 */
	koord3d pos;

	/**
	 * x-offset of the object on the tile
	 * used for drawing object image
	 * if xoff and yoff are between 0 and OBJECT_OFFSET_STEPS-1 then
	 * the top-left corner of the image is within tile boundaries
	 */
	sint8 xoff;

	/**
	 * y-offset of the object on the tile
	 */
	sint8 yoff;

	/**
	 * Owner of the object (1 - public player, 15 - unowned)
	 */
	uint8 owner_n:4;

	/**
	 * @see flag_values
	 */
	uint8 flags:4;

private:
	/**
	* Used by all constructors to initialize all vars with safe values
	* -> single source principle
	* @author Hj. Malthaner
	*/
	void init();

protected:
	obj_t();

	// since we often need access during loading
	void set_player_nr(uint8 o) { owner_n = o; }

	/**
	* Pointer to the world of this thing. Static to conserve space.
	* Change to instance variable once more than one world is available.
	* @author Hj. Malthaner
	*/
	static karte_ptr_t welt;


public:
	// needed for drawing images
	sint8 get_player_nr() const { return owner_n; }

	/**
	 * sets owner of object
	 */
	void set_owner(player_t *player);

	/**
	 * returns owner of object
	 */
	player_t * get_owner() const;

	/**
	 * routines to set, clear, get bit flags
	 * @author Hj. Malthaner
	 */
	inline void set_flag(flag_values flag) {flags |= flag;}
	inline void clear_flag(flag_values flag) {flags &= ~flag;}
	inline bool get_flag(flag_values flag) const {return ((flags & flag) != 0);}

	/// all the different types of objects
	enum typ {
		undefined=-1, obj=0, baum=1, zeiger=2,
		wolke=3, sync_wolke=4, async_wolke=5,

		gebaeude=7,	// animated buildings (6 not used any more)
		signal=8,

		bruecke=9, tunnel=10,
		bahndepot=12, strassendepot=13, schiffdepot = 14,

		raucher=15, // obsolete
		leitung = 16, pumpe = 17, senke = 18,
		roadsign = 19, pillar = 20,

		airdepot = 21, monoraildepot=22, tramdepot=23, maglevdepot=24,

		wayobj = 25,
		way = 26, // since 99.04 ways are normal things and stored in the objliste_t!

		label = 27,	// indicates ownership
		field = 28,
		crossing = 29,
		groundobj = 30, // lakes, stones

		narrowgaugedepot=31,

		// after this only moving stuff
		// reserved values for vehicles: 64 to 95
		pedestrian=64,
		road_user=65,
		road_vehicle=66,
		rail_vehicle=67,
		monorail_vehicle=68,
		maglev_vehicle=69,
		narrowgauge_vehicle=70,
		water_vehicle=80,
		air_vehicle=81,
		movingobj=82,

		// other new objs (obsolete, only used during loading old games
		// lagerhaus = 24, (never really used)
		// gebaeude_alt=6,	(very, very old?)
		old_gebaeudefundament=11,	// wall below buildings, not used any more
		old_automobil=32, old_waggon=33,
		old_schiff=34, old_aircraft=35, old_monorailwaggon=36,
		old_verkehr=41,
		old_fussgaenger=42,
		old_choosesignal = 95,
		old_presignal = 96,
		old_roadsign = 97,
		old_pillar = 98,
		old_airdepot = 99,
		old_monoraildepot=100,
		old_tramdepot=101,
	};

	inline sint8 get_xoff() const {return xoff;}
	inline sint8 get_yoff() const {return yoff;}

	// true for all moving objects
	inline bool is_moving() const { return flags&is_vehicle; }

	// while in principle, this should trigger the dirty, it takes just too much time to do it
	// TAKE CARE OF SET IT DIRTY YOURSELF!!!
	inline void set_xoff(sint8 xoff) {this->xoff = xoff; }
	inline void set_yoff(sint8 yoff) {this->yoff = yoff; }

	/**
	 * Constructor to load object from file
	 * @author Hj. Malthaner
	 */
	obj_t(loadsave_t *file);

	/**
	 * Constructor to set position of object
	 * This does *not* add the object to the tile
	 * @author Hj. Malthaner
	 */
	obj_t(koord3d pos);

	/**
	 * Destructor: removes object from tile, should close any inspection windows
	 * @author Hj. Malthaner
	 */
	virtual ~obj_t();

	/**
	 * Routine for cleanup if object is removed (ie book maintenance, cost for removal)
	 * @author Hj. Malthaner
	 */
	virtual void cleanup(player_t *) {}

	/**
	 * @returns untranslated name of object
	 * @author Hj. Malthaner
	 */
	virtual const char *get_name() const {return "Ding";}

	/**
	 * @return object type
	 * @author Hj. Malthaner
	 * @see typ
	 */
	virtual typ get_typ() const = 0;

	/**
	 * waytype associated with this object
	 */
	virtual waytype_t get_waytype() const { return invalid_wt; }

	/**
	 * called whenever the snowline height changes
	 * return false and the obj_t will be deleted
	 * @author prissi
	 */
	virtual bool check_season(const bool) { return true; }

	/**
	 * called during map rotation
	 * @author priss
	 */
	virtual void rotate90();

	/**
	 * Every object needs an image.
	 * @return number of current image for that object
	 * @author Hj. Malthaner
	 */
	virtual image_id get_image() const = 0;

	/**
	 * give image for height > 0 (max. height currently 3)
	 * IMG_EMPTY is no images
	 * @author Hj. Malthaner
	 */
	virtual image_id get_image(int /*height*/) const {return IMG_EMPTY;}

	/**
	 * this image is drawn after all get_image() on this tile
	 * Currently only single height is supported for this feature
	 */
	virtual image_id get_front_image() const {return IMG_EMPTY;}

	/**
	 * if a function returns a value here with TRANSPARENT_FLAGS set
	 * then a transparent outline with the color from the lower 8 bit is drawn
	 * @author kierongreen
	 */
	virtual FLAGGED_PIXVAL get_outline_colour() const {return 0;}

	/**
	 * The image, that will be outlined
	 * @author kierongreen
	 */
	virtual image_id get_outline_image() const { return IMG_EMPTY; }

	/**
	 * Save and Load of object data in one routine
	 * @author Hj. Malthaner
	 */
	virtual void rdwr(loadsave_t *file);

	/**
	 * Called after the world is completely loaded from savegame
	 *
	 * @author Hj. Malthaner
	 */
	virtual void finish_rd() {}

	/**
	 * @return position
	 */
	inline koord3d get_pos() const {return pos;}

	/**
	 * set position - you would not have guessed it :)
	 */
	inline void set_pos(koord3d k) { if(k!=pos) { set_flag(dirty); pos = k;} }

	/**
	 * put description of object into the buffer
	 * (used for certain windows)
	 * @author Hj. Malthaner
	 * @see simwin
	 */
	virtual void info(cbuffer_t & buf) const;

	/**
	 * Opens a new info window for the object
	 * @author Hj. Malthaner
	 */
	virtual void show_info();

	/**
	 * @return NULL if OK, otherwise an error message
	 * @author Hj. Malthaner
	 */
	virtual const char *is_deletable(const player_t *player);

	/**
	 * Draw background image of object
	 * (everything that could be potentially behind vehicles)
	 */
	void display(int xpos, int ypos  CLIP_NUM_DEF) const;

	/**
	 * Draw foreground image
	 * (everything that is in front of vehicles)
	 */
#ifdef MULTI_THREAD
	virtual void display_after(int xpos, int ypos, const sint8 clip_num) const;
#else
	virtual void display_after(int xpos, int ypos, bool is_global) const;
#endif

#ifdef MULTI_THREAD
	/**
	 * Draw overlays
	 * (convoi tooltips)
	 */
	virtual void display_overlay(int /*xpos*/, int /*ypos*/) const { return; }
#endif

	/**
	* When a vehicle moves or a cloud moves, it needs to mark the old spot as dirty (to copy to screen).
	* This routine already takes position, and offsets (x_off, y_off) into account.
	* @param yoff extra y-offset, in most cases 0, in pixels.
	* @author prissi
	*/
	void mark_image_dirty(image_id image, sint16 yoff) const;

	/**
	 * Function for recalculating the image.
	 * @author Hj. Malthaner
	 */
	virtual void calc_image() {}
};


/**
 * Template to do casting of pointers based on obj_t::typ
 * as a replacement of the slower dynamic_cast<>
 */
template<typename T> static T* obj_cast(obj_t*);

template<typename T> static inline T const* obj_cast(obj_t const* const d)
{
	return obj_cast<T>(const_cast<obj_t*>(d));
}


/**
 * Game objects that do not have description windows (for instance zeiger_t, wolke_t)
 */
class obj_no_info_t : public obj_t
{
public:
	obj_no_info_t(loadsave_t* file) : obj_t(file) {}

	obj_no_info_t(koord3d pos) : obj_t(pos) {}

	void show_info() {}

protected:
	obj_no_info_t() : obj_t() {}
};

#endif
